EIPS/eip-2315.md

```diff
@@ -0,0 +1,78 @@
+---
+eip: 2315
+title: Simple Subroutines for the EVM
+status: Draft
+type: Standards Track
+category: Core
+author: Greg Colvin
+discussions-to: https://ethereum-magicians.org/t/eip-2315-simple-subroutines-for-the-evm/3941
+created: 2019-10-17
+---
+## Abstract
+
+This proposal introduces two opcodes to support subroutines:  `JUMPSUB` and `RETSUB`.
+
+## Motivation
+
+The EVM does not provide subroutines as a primitive.  Instead, calls must be synthesized by fetching and pushing the current program counter on the data stack and jumping to the subroutine address; returns must be synthesized by getting the return address to the top of stack and jumping back to it.
+
+## Specification
+
+##### `JUMPSUB`
+Jumps to the address on top of the stack, which must be the offset of a `JUMPDEST`.
+
+##### `RETSUB`
+Returns to the most recently executed `JUMPSUB` instruction.
+
+## Rationale
+
+This is the smallest possible change that provides native subroutines without breaking backwards compatibility.
+
+## Backwards Compatibility
+
+These changes do not affect the semantics of existing EVM code.
+
+## Test Cases
+```
+step op       stack
+0    PUSH1 3  []
+1    JUMPSUB  [3]
+4    STOP
+2    JUMPDEST []
+3    RETSUB   []
+```
```

Comment:
This is how I interpret it. 
Program:
```
pc op
0    PUSH1 3
2    JUMPSUB
3    JUMPDEST
4    RETSUB
```
Trace:
```
step pc op       stack
0    0   PUSH1 3  []
1    2   JUMPSUB  [3]
2    3   JUMPDEST []
3    4  RETSUB []
4    2 JUMPSUB [] // Exception due to shallow stack
```

Reply:
see update

Reply:
No, now I'm confused again. The `RETSUB` would return to `pc=2`, the `JUMPSUB`. And that one would hit shallow stack, but not virtual `STOP`. So it would end execution on error, right?

Reply:
Just woke up. other duties.  I might not be possible not to break something here.  But then I doubt we have a list of every abstraction and assumption made by the current VM.

Reply:
Three cups of coffee.  Getting closer?
```
step pc op       stack
0    1 PUSH1 4  []
1    2 JUMPSUB  [3]
2    4 JUMPDEST []
3    5 RETSUB   []
4    2 JUMPSUB  []
5    3 STOP
```
Notes:
* at step 0 PUSH1 advances to 1 = ++PC and push(data_stack, 4)
* at step 1 JUMPSUB jumps to 4 = PC = pop(data_stack), and push(return_stack, 2)
* at step 2 JUMPDEST advances to 5 = ++PC
* at step 3 RETSUB returns to 2 = PC = pop(return_stack)
* at step 4 JUMPSUB advances to 5 = ++PC
* at step 5 program STOPs with empty stack

Reply:
I think what is confusing here is that in the typical implementation the ++PC is done at the end of the loop.  So it's not exactly the case that JUMPSUB does that at step 4, although it is the case that RETSUB pops the return stack to PC at step3.  It might be less confusing to say that RETSUB returns to 3 == PC = pop(return_stack)  + 1, and adust the text and pseudocode to match.

Reply:
Your `JUMPSUB [3]` should be `JUMPSUB[4]` or your `PUSH1 4` needs a change. Could you also provide the program bytecode for that trace? 

The spec for `JUMPSUB` is

> Jumps to the address on top of the stack, which must be the offset of a `JUMPDEST`. 

It does not mention any special casing of how empty stacks are handled, so IMO that means it errors out. The only time you mention a special casing is for the `RETSUB`: 

> The virtual byte of 0 at this offset is the EVM `STOP` opcode, so executing a `RETSUB` with no prior `JUMPSUB` executes a `STOP`.  A `STOP` or `RETURN` ends the execution of the subroutine and the program.

>  It might be less confusing to say that RETSUB returns to 3 == PC = pop(return_stack) + 1

Yes, if the intention is to continue _after_ the 'current' `JUMPSUB`, then that makes a lot more sense. 

For extra clarity, I'd suggest making the program maybe 

```
pc   op
0    PUSH1 5
2    JUMPSUB
3    CALLDATA
4    STOP
5    JUMPDEST
6    PUSH1 01
8    POP
9    RETSUB
```

IIUC, the trace would then go through PCs `0,2,5,6,8,9,3,4`. Is that the intention?